vfs: flaky test TestDiskHealthChecking_Filesystem

See: https://github.com/cockroachdb/pebble/runs/6563748979?check_suite_focus=true

```console
--- FAIL: TestDiskHealthChecking_Filesystem (0.41s)
    --- FAIL: TestDiskHealthChecking_Filesystem/remove-all (0.05s)
        assertion_compare.go:211: 
            	Error Trace:	disk_health_test.go:276
            	Error:      	"0" is not greater than "0"
            	Test:       	TestDiskHealthChecking_Filesystem/remove-all
            	Messages:   	[]
```

Another one on macOS: https://github.com/cockroachdb/pebble/runs/7083257581?check_suite_focus=true

```console
--- FAIL: TestDiskHealthChecking_Filesystem (0.44s)
    --- FAIL: TestDiskHealthChecking_Filesystem/remove (0.05s)
        assertion_compare.go:313: 
            	Error Trace:	disk_health_test.go:276
            	Error:      	"0" is not greater than "0"
            	Test:       	TestDiskHealthChecking_Filesystem/remove
            	Messages:   	[]
```

Managed to reproduce this on my macbook. Looks like the same run caught another test failure:

```bash
$ go test -mod=vendor -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run TestDiskHealthChecking_Filesystem ./vfs
...
/var/folders/ps/wznnm1sn4_g1rn8jjlgx57rh0000gq/T/go-stress-20220627T170345-028551772
=== RUN   TestDiskHealthChecking_Filesystem
=== RUN   TestDiskHealthChecking_Filesystem/rename
=== RUN   TestDiskHealthChecking_Filesystem/reuse-for-write
=== RUN   TestDiskHealthChecking_Filesystem/create
=== RUN   TestDiskHealthChecking_Filesystem/link
=== RUN   TestDiskHealthChecking_Filesystem/mkdir-all
    assertion_compare.go:313:
                Error Trace:    disk_health_test.go:276
                Error:          "0" is not greater than "0"
                Test:           TestDiskHealthChecking_Filesystem/mkdir-all
                Messages:       []
=== RUN   TestDiskHealthChecking_Filesystem/remove
    assertion_compare.go:313:
                Error Trace:    disk_health_test.go:276
                Error:          "0" is not greater than "0"
                Test:           TestDiskHealthChecking_Filesystem/remove
                Messages:       []
=== RUN   TestDiskHealthChecking_Filesystem/remove-all
--- FAIL: TestDiskHealthChecking_Filesystem (0.56s)
    --- PASS: TestDiskHealthChecking_Filesystem/rename (0.11s)
    --- PASS: TestDiskHealthChecking_Filesystem/reuse-for-write (0.05s)
    --- PASS: TestDiskHealthChecking_Filesystem/create (0.08s)
    --- PASS: TestDiskHealthChecking_Filesystem/link (0.08s)
    --- FAIL: TestDiskHealthChecking_Filesystem/mkdir-all (0.07s)
    --- FAIL: TestDiskHealthChecking_Filesystem/remove (0.07s)
    --- PASS: TestDiskHealthChecking_Filesystem/remove-all (0.05s)
=== RUN   TestDiskHealthChecking_Filesystem_Close
--- PASS: TestDiskHealthChecking_Filesystem_Close (0.15s)
FAIL


ERROR: exit status 1

34m5s: 3902 runs so far, 2 failures (0.05%)
```

This doesn't happen consistently / frequently enough to warrant too much investigation right now, but it's good to know that this is somewhat reproducible.

Haven't seen this one before:

```console
--- FAIL: TestDiskHealthChecking_Filesystem_Close (0.17s)
    disk_health_test.go:310: 
        	Error Trace:	disk_health_test.go:310
        	Error:      	map[string]time.Duration{"bar":40320000, "foo":56489000} does not contain "bax"
        	Test:       	TestDiskHealthChecking_Filesystem_Close
```

--- a/vfs/disk_health_test.go
+++ b/vfs/disk_health_test.go
@@ -8,6 +8,8 @@ import (
 	"io"
 	"math"
 	"os"
+	"runtime"
+	"sync"
 	"sync/atomic"
 	"testing"
 	"time"
@@ -216,6 +218,10 @@ var _ FS = &mockFS{}
 func TestDiskHealthChecking_File(t *testing.T) {
 	oldTickInterval := defaultTickInterval
 	defaultTickInterval = time.Millisecond
+	if runtime.GOOS == "windows" {
+		t.Skipf("skipped on windows due to unreliable runtimes")
+	}
+
 	defer func() { defaultTickInterval = oldTickInterval }()
 
 	const (
@@ -229,43 +235,23 @@ func TestDiskHealthChecking_File(t *testing.T) {
 		writeDuration    time.Duration
 		fn               func(f File)
 		createWriteDelta time.Duration
-		expectStall      bool
 	}{
-		// No false negatives.
 		{
 			op:            OpTypeWrite,
 			writeSize:     5 * writeSizePrecision, // five KB
 			writeDuration: 100 * time.Millisecond,
 			fn:            func(f File) { f.Write(fiveKB) },
-			expectStall:   true,
 		},
 		{
 			op:            OpTypeSync,
 			writeSize:     0,
 			writeDuration: 100 * time.Millisecond,
 			fn:            func(f File) { f.Sync() },
-			expectStall:   true,
-		},
-		// No false positives.
-		{
-			op:               OpTypeWrite,
-			writeSize:        5,
-			writeDuration:    25 * time.Millisecond,
-			fn:               func(f File) { f.Write([]byte("uh oh")) },
-			createWriteDelta: 100 * time.Millisecond,
-			expectStall:      false,
-		},
-		{
-			op:            OpTypeSync,
-			writeSize:     0,
-			writeDuration: 25 * time.Millisecond,
-			fn:            func(f File) { f.Sync() },
-			expectStall:   false,
 		},
 	}
 	for _, tc := range testCases {
 		t.Run(tc.op.String(), func(t *testing.T) {
-			diskSlow := make(chan DiskSlowInfo, 1)
+			diskSlow := make(chan DiskSlowInfo, 3)
 			mockFS := &mockFS{create: func(name string) (File, error) {
 				return mockFile{syncAndWriteDuration: tc.writeDuration}, nil
 			}}
@@ -283,25 +269,16 @@ func TestDiskHealthChecking_File(t *testing.T) {
 
 			tc.fn(dhFile)
 
-			if tc.expectStall { // no false negatives
-				select {
-				case i := <-diskSlow:
-					d := i.Duration
-					if d.Seconds() < slowThreshold.Seconds() {
-						t.Fatalf("expected %0.1f to be greater than threshold %0.1f", d.Seconds(), slowThreshold.Seconds())
-					}
-					require.Equal(t, tc.writeSize, i.WriteSize)
-					require.Equal(t, tc.op, i.OpType)
-				case <-time.After(200 * time.Millisecond):
-					t.Fatal("disk stall detector did not detect slow disk operation")
+			select {
+			case i := <-diskSlow:
+				d := i.Duration
+				if d.Seconds() < slowThreshold.Seconds() {
+					t.Fatalf("expected %0.1f to be greater than threshold %0.1f", d.Seconds(), slowThreshold.Seconds())
 				}
-			} else { // no false positives
-				select {
-				case <-diskSlow:
-					t.Fatal("disk stall detector detected a slow disk operation")
-				case <-time.After(200 * time.Millisecond):
-					return
-				}
+				require.Equal(t, tc.writeSize, i.WriteSize)
+				require.Equal(t, tc.op, i.OpType)
+			case <-time.After(10 * time.Second):
+				t.Fatal("disk stall detector did not detect slow disk operation")
 			}
 		})
 	}
@@ -457,7 +434,12 @@ func filesystemOpsMockFS(sleepDur time.Duration) *mockFS {
 func stallFilesystemOperations(fs FS) []filesystemOperation {
 	return []filesystemOperation{
 		{
-			"create", OpTypeCreate, func() { _, _ = fs.Create("foo") },
+			"create", OpTypeCreate, func() {
+				f, _ := fs.Create("foo")
+				if f != nil {
+					f.Close()
+				}
+			},
 		},
 		{
 			"link", OpTypeLink, func() { _ = fs.Link("foo", "bar") },
@@ -489,15 +471,24 @@ type filesystemOperation struct {
 func TestDiskHealthChecking_Filesystem(t *testing.T) {
 	const sleepDur = 50 * time.Millisecond
 	const stallThreshold = 10 * time.Millisecond
+	if runtime.GOOS == "windows" {
+		t.Skipf("skipped on windows due to unreliable runtimes")
+	}
 
 	// Wrap with disk-health checking, counting each stall via stallCount.
 	var expectedOpType OpType
 	var stallCount atomic.Uint64
+	onStall := make(chan struct{}, 10)
+	var lastOpType OpType
 	fs, closer := WithDiskHealthChecks(filesystemOpsMockFS(sleepDur), stallThreshold,
 		func(info DiskSlowInfo) {
 			require.Equal(t, 0, info.WriteSize)
-			require.Equal(t, expectedOpType, info.OpType)
 			stallCount.Add(1)
+			if lastOpType != info.OpType {
+				require.Equal(t, expectedOpType, info.OpType)
+				lastOpType = info.OpType
+				onStall <- struct{}{}
+			}
 		})
 	defer closer.Close()
 	fs.(*diskHealthCheckingFS).tickInterval = 5 * time.Millisecond
@@ -507,6 +498,11 @@ func TestDiskHealthChecking_Filesystem(t *testing.T) {
 			expectedOpType = o.opType
 			before := stallCount.Load()
 			o.f()
+			select {
+			case <-onStall:
+			case <-time.After(10 * time.Second):
+				t.Fatal("timed out waiting for stall")
+			}
 			after := stallCount.Load()
 			require.Greater(t, int(after-before), 0)
 		})
@@ -521,27 +517,53 @@ func TestDiskHealthChecking_Filesystem(t *testing.T) {
 // multiple times.
 func TestDiskHealthChecking_Filesystem_Close(t *testing.T) {
 	const stallThreshold = 10 * time.Millisecond
+	stallChan := make(chan struct{}, 1)
 	mockFS := &mockFS{
 		create: func(name string) (File, error) {
-			time.Sleep(50 * time.Millisecond)
+			<-stallChan
 			return &mockFile{}, nil
 		},
 	}
 
-	stalled := map[string]time.Duration{}
+	files := []string{"foo", "bar", "bax"}
+	var lastPath string
+	stalled := make(chan string)
 	fs, closer := WithDiskHealthChecks(mockFS, stallThreshold,
-		func(info DiskSlowInfo) { stalled[info.Path] = info.Duration })
+		func(info DiskSlowInfo) {
+			if lastPath != info.Path {
+				lastPath = info.Path
+				stalled <- info.Path
+			}
+		})
 	fs.(*diskHealthCheckingFS).tickInterval = 5 * time.Millisecond
 
-	files := []string{"foo", "bar", "bax"}
+	var wg sync.WaitGroup
 	for _, filename := range files {
-		// Create will stall, and the detector should write to the stalled map
+		filename := filename
+		// Create will stall, and the detector should write to the stalled channel
 		// with the filename.
-		_, _ = fs.Create(filename)
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			f, _ := fs.Create(filename)
+			if f != nil {
+				f.Close()
+			}
+		}()
+
+		select {
+		case stalledPath := <-stalled:
+			require.Equal(t, filename, stalledPath)
+		case <-time.After(10 * time.Second):
+			t.Fatalf("timed out waiting for stall")
+		}
+		// Unblock the call to Create().
+		stallChan <- struct{}{}
+
 		// Invoke the closer. This will cause the long-running goroutine to
 		// exit, but the fs should still be usable and should still detect
 		// subsequent stalls on the next iteration.
 		require.NoError(t, closer.Close())
-		require.Contains(t, stalled, filename)
 	}
+	wg.Wait()
 }

